File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/AbsAttachmentsEditPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import static dev.ragnarok.fenrir.util.AppPerms.hasCameraPermission;
import static dev.ragnarok.fenrir.util.AppPerms.hasReadStoragePermission;
import static dev.ragnarok.fenrir.util.Objects.isNull;
import static dev.ragnarok.fenrir.util.Objects.nonNull;
import static dev.ragnarok.fenrir.util.Utils.findInfoByPredicate;
import static dev.ragnarok.fenrir.util.Utils.nonEmpty;
import static dev.ragnarok.fenrir.util.Utils.safeCountOfMultiple;

import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Includes;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.model.AbsModel;
import dev.ragnarok.fenrir.model.AttachmentEntry;
import dev.ragnarok.fenrir.model.LocalPhoto;
import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.model.Poll;
import dev.ragnarok.fenrir.mvp.presenter.base.AccountDependencyPresenter;
import dev.ragnarok.fenrir.mvp.view.IBaseAttachmentsEditView;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.upload.IUploadManager;
import dev.ragnarok.fenrir.upload.Upload;
import dev.ragnarok.fenrir.util.FileUtil;
import dev.ragnarok.fenrir.util.Pair;
import dev.ragnarok.fenrir.util.Predicate;

public abstract class AbsAttachmentsEditPresenter<V extends IBaseAttachmentsEditView>
        extends AccountDependencyPresenter<V> {

    private static final String SAVE_DATA = "save_data";
    private static final String SAVE_TIMER = "save_timer";
    private static final String SAVE_BODY = "save_body";
    private static final String SAVE_CURRENT_PHOTO_CAMERA_URI = "save_current_photo_camera_uri";
    final IUploadManager uploadManager;
    private final ArrayList<AttachmentEntry> data;
    private String textBody;
    private Uri currentPhotoCameraUri;
    private Long timerValue;

    AbsAttachmentsEditPresenter(int accountId, @Nullable Bundle savedInstanceState) {
        super(accountId, savedInstanceState);
        uploadManager = Includes.getUploadManager();

        if (nonNull(savedInstanceState)) {
            currentPhotoCameraUri = savedInstanceState.getParcelable(SAVE_CURRENT_PHOTO_CAMERA_URI);
            textBody = savedInstanceState.getString(SAVE_BODY);
            timerValue = savedInstanceState.containsKey(SAVE_TIMER) ? savedInstanceState.getLong(SAVE_TIMER) : null;
        }

        data = new ArrayList<>();
        if (nonNull(savedInstanceState)) {
            ArrayList<AttachmentEntry> savedEntries = savedInstanceState.getParcelableArrayList(SAVE_DATA);
            if (nonEmpty(savedEntries)) {
                data.addAll(savedEntries);
            }
        }
    }

    static List<AttachmentEntry> createFrom(List<Upload> objects) {
        List<AttachmentEntry> data = new ArrayList<>(objects.size());
        for (Upload object : objects) {
            data.add(new AttachmentEntry(true, object));
        }

        return data;
    }

    static List<AttachmentEntry> createFrom(List<Pair<Integer, AbsModel>> pairs, boolean canDelete) {
        List<AttachmentEntry> data = new ArrayList<>(pairs.size());
        for (Pair<Integer, AbsModel> pair : pairs) {
            data.add(new AttachmentEntry(canDelete, pair.getSecond()).setOptionalId(pair.getFirst()));
        }
        return data;
    }

    Long getTimerValue() {
        return timerValue;
    }

    void setTimerValue(Long timerValue) {
        this.timerValue = timerValue;
        resolveTimerInfoView();
    }

    ArrayList<AttachmentEntry> getData() {
        return data;
    }

    void resolveTimerInfoView() {
        callView(v -> v.setTimerValue(timerValue));
    }

    void resolveTextView() {
        callView(v -> v.setTextBody(textBody));
    }

    String getTextBody() {
        return textBody;
    }

    void setTextBody(String body) {
        textBody = body;
        resolveTextView();
    }

    ArrayList<AttachmentEntry> getNeedParcelSavingEntries() {
        return new ArrayList<>(0);
    }

    @Override
    public void saveState(@NonNull Bundle outState) {
        super.saveState(outState);
        outState.putParcelable(SAVE_CURRENT_PHOTO_CAMERA_URI, currentPhotoCameraUri);
        outState.putParcelableArrayList(SAVE_DATA, getNeedParcelSavingEntries());
        outState.putString(SAVE_BODY, textBody);
        if (nonNull(timerValue)) {
            outState.putLong(SAVE_TIMER, timerValue);
        }
    }

    void onUploadProgressUpdate(List<IUploadManager.IProgressUpdate> updates) {
        for (IUploadManager.IProgressUpdate update : updates) {
            Predicate<AttachmentEntry> predicate = entry -> entry.getAttachment() instanceof Upload
                    && ((Upload) entry.getAttachment()).getId() == update.getId();

            Pair<Integer, AttachmentEntry> info = findInfoByPredicate(getData(), predicate);

            if (nonNull(info)) {
                AttachmentEntry entry = info.getSecond();

                Upload object = (Upload) entry.getAttachment();
                if (object.getStatus() != Upload.STATUS_UPLOADING) {
                    continue;
                }

                callView(v -> v.updateProgressAtIndex(entry.getId(), update.getProgress()));
            }
        }
    }

    void onUploadObjectRemovedFromQueue(int[] ids) {
        for (int id : ids) {
            int index = findUploadIndexById(id);
            if (index != -1) {
                manuallyRemoveElement(index);
            }
        }
    }

    void onUploadQueueUpdates(List<Upload> updates, Predicate<Upload> predicate) {
        int startSize = data.size();
        int count = 0;

        for (Upload u : updates) {
            if (predicate.test(u)) {
                data.add(new AttachmentEntry(true, u));
                count++;
            }
        }

        if (count > 0) {
            safelyNotifyItemsAdded(startSize, count);
        }
    }

    void safelyNotifyItemsAdded(int position, int count) {
        callView(v -> v.notifyItemRangeInsert(position, count));
    }

    List<AttachmentEntry> combine(List<AttachmentEntry> first, List<AttachmentEntry> second) {
        List<AttachmentEntry> data = new ArrayList<>(safeCountOfMultiple(first, second));
        data.addAll(first);
        data.addAll(second);
        return data;
    }

    void onUploadStatusUpdate(Upload update) {
        int index = findUploadIndexById(update.getId());
        if (index != -1) {
            safeNotifyDataSetChanged();
        }
    }

    @Override
    public void onGuiCreated(@NonNull V viewHost) {
        super.onGuiCreated(viewHost);
        viewHost.displayInitialModels(data);

        resolveTimerInfoView();
        resolveTextView();
    }

    public final void fireRemoveClick(int index, @NonNull AttachmentEntry attachment) {
        if (attachment.getAttachment() instanceof Upload) {
            Upload upload = (Upload) attachment.getAttachment();
            uploadManager.cancel(upload.getId());
            return;
        }

        onAttachmentRemoveClick(index, attachment);
    }

    void safelyNotifyItemRemoved(int position) {
        callView(v -> v.notifyItemRemoved(position));
    }

    void onAttachmentRemoveClick(int index, @NonNull AttachmentEntry attachment) {
        throw new UnsupportedOperationException();
    }

    void manuallyRemoveElement(int index) {
        data.remove(index);
        safelyNotifyItemRemoved(index);

        //safeNotifyDataSetChanged();
    }

    public final void fireTitleClick(int index, @NonNull AttachmentEntry attachment) {

    }

    private int getMaxCountOfAttachments() {
        return 10;
    }

    private boolean canAttachMore() {
        return data.size() < getMaxCountOfAttachments();
    }

    private int getMaxFutureAttachmentCount() {
        int count = data.size() - getMaxCountOfAttachments();
        return Math.max(count, 0);
    }

    public final void firePhotoFromVkChoose() {
        callView(v -> v.openAddVkPhotosWindow(getMaxFutureAttachmentCount(), getAccountId(), getAccountId()));
    }

    private boolean checkAbilityToAttachMore() {
        if (canAttachMore()) {
            return true;
        } else {
            callView(v -> v.showError(R.string.reached_maximum_count_of_attachments));
            return false;
        }
    }

    public final void firePhotoFromLocalGalleryChoose() {
        if (!hasReadStoragePermission(getApplicationContext())) {
            callView(IBaseAttachmentsEditView::requestReadExternalStoragePermission);
            return;
        }

        callView(v -> v.openAddPhotoFromGalleryWindow(getMaxFutureAttachmentCount()));
    }

    public final void firePhotoFromCameraChoose() {
        if (!hasCameraPermission(getApplicationContext())) {
            callView(IBaseAttachmentsEditView::requestCameraPermission);
            return;
        }

        createImageFromCamera();
    }

    private void createImageFromCamera() {
        try {
            File photoFile = FileUtil.createImageFile();
            currentPhotoCameraUri = FileUtil.getExportedUriForFile(getApplicationContext(), photoFile);

            callView(v -> v.openCamera(currentPhotoCameraUri));
        } catch (IOException e) {
            callView(v -> v.showError(e.getMessage()));
        }
    }

    public final void firePhotoMaked() {
        callView(v -> v.notifySystemAboutNewPhoto(currentPhotoCameraUri));

        LocalPhoto makedPhoto = new LocalPhoto().setFullImageUri(currentPhotoCameraUri);
        doUploadPhotos(Collections.singletonList(makedPhoto));
    }

    protected void doUploadPhotos(List<LocalPhoto> photos, int size) {
        throw new UnsupportedOperationException();
    }

    protected void doUploadFile(String file, int size) {
        throw new UnsupportedOperationException();
    }

    public final void fireFileSelected(String file) {
        doUploadFile(file, Upload.IMAGE_SIZE_FULL);
    }

    private void doFinalUploadPhotos(List<LocalPhoto> photos, int size) {
        if (size == Upload.IMAGE_SIZE_CROPPING && photos.size() == 1) {
            Uri to_up = photos.get(0).getFullImageUri();
            if (new File(to_up.getPath()).isFile()) {
                to_up = Uri.fromFile(new File(to_up.getPath()));
            }
            Uri finalTo_up = to_up;
            callView(v -> v.displayCropPhotoDialog(finalTo_up));
        } else {
            doUploadPhotos(photos, size);
        }
    }

    private void doUploadPhotos(List<LocalPhoto> photos) {
        Integer size = Settings.get()
                .main()
                .getUploadImageSize();

        if (isNull(size)) {
            callView(v -> v.displaySelectUploadPhotoSizeDialog(photos));
        } else {
            doFinalUploadPhotos(photos, size);
        }
    }

    public final void firePhotosFromGallerySelected(ArrayList<LocalPhoto> photos) {
        doUploadPhotos(photos);
    }

    public final void fireButtonPhotoClick() {
        if (checkAbilityToAttachMore()) {
            callView(IBaseAttachmentsEditView::displayChoosePhotoTypeDialog);
        }
    }

    public final void fireButtonAudioClick() {
        if (checkAbilityToAttachMore()) {
            callView(v -> v.openAddAudiosWindow(getMaxFutureAttachmentCount(), getAccountId()));
        }
    }

    public final void fireButtonVideoClick() {
        if (checkAbilityToAttachMore()) {
            callView(v -> v.openAddVideosWindow(getMaxFutureAttachmentCount(), getAccountId()));
        }
    }

    public final void fireButtonDocClick() {
        if (checkAbilityToAttachMore()) {
            callView(v -> v.openAddDocumentsWindow(getMaxFutureAttachmentCount(), getAccountId()));
        }
    }

    protected void onPollCreateClick() {
        throw new UnsupportedOperationException();
    }

    protected void onTimerClick() {
        throw new UnsupportedOperationException();
    }

    public final void fireButtonPollClick() {
        onPollCreateClick();
    }

    public final void fireButtonTimerClick() {
        onTimerClick();
    }

    protected void onModelsAdded(List<? extends AbsModel> models) {
        for (AbsModel model : models) {
            data.add(new AttachmentEntry(true, model));
        }

        safeNotifyDataSetChanged();
    }

    public final void fireAttachmentsSelected(@NonNull ArrayList<? extends AbsModel> attachments) {
        onModelsAdded(attachments);
    }

    public void fireUploadPhotoSizeSelected(@NonNull List<LocalPhoto> photos, int size) {
        doFinalUploadPhotos(photos, size);
    }

    public final void firePollCreated(@NonNull Poll poll) {
        onModelsAdded(Collections.singletonList(poll));
    }

    protected void safeNotifyDataSetChanged() {
        callView(IBaseAttachmentsEditView::notifyDataSetChanged);
    }

    public final void fireTextChanged(CharSequence s) {
        textBody = isNull(s) ? null : s.toString();
    }

    public final void fireVkPhotosSelected(@NonNull ArrayList<Photo> photos) {
        onModelsAdded(photos);
    }

    public void fireCameraPermissionResolved() {
        if (hasCameraPermission(getApplicationContext())) {
            createImageFromCamera();
        }
    }

    public void fireReadStoragePermissionResolved() {
        if (hasReadStoragePermission(getApplicationContext())) {
            callView(v -> v.openAddPhotoFromGalleryWindow(getMaxFutureAttachmentCount()));
        }
    }

    boolean hasUploads() {
        for (AttachmentEntry entry : data) {
            if (entry.getAttachment() instanceof Upload) {
                return true;
            }
        }

        return false;
    }

    int findUploadIndexById(int id) {
        for (int i = 0; i < data.size(); i++) {
            AttachmentEntry item = data.get(i);
            if (item.getAttachment() instanceof Upload && ((Upload) item.getAttachment()).getId() == id) {
                return i;
            }
        }

        return -1;
    }

    public void fireTimerTimeSelected(long unixtime) {
        throw new UnsupportedOperationException();
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/attachments/absattachmentsedit/AbsAttachmentsEditPresenter.kt
package dev.ragnarok.fenrir.fragment.attachments.absattachmentsedit

import android.net.Uri
import android.os.Bundle
import dev.ragnarok.fenrir.*
import dev.ragnarok.fenrir.fragment.base.AccountDependencyPresenter
import dev.ragnarok.fenrir.model.*
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.upload.IUploadManager
import dev.ragnarok.fenrir.upload.IUploadManager.IProgressUpdate
import dev.ragnarok.fenrir.upload.Upload
import dev.ragnarok.fenrir.util.AppPerms.hasCameraPermission
import dev.ragnarok.fenrir.util.AppPerms.hasReadStoragePermission
import dev.ragnarok.fenrir.util.FileUtil.createImageFile
import dev.ragnarok.fenrir.util.FileUtil.getExportedUriForFile
import dev.ragnarok.fenrir.util.Pair
import dev.ragnarok.fenrir.util.Utils.findInfoByPredicate
import dev.ragnarok.fenrir.util.Utils.safeCountOfMultiple
import java.io.File
import java.io.IOException

abstract class AbsAttachmentsEditPresenter<V : IBaseAttachmentsEditView> internal constructor(
    accountId: Int,
    savedInstanceState: Bundle?
) : AccountDependencyPresenter<V>(accountId, savedInstanceState) {
    val uploadManager: IUploadManager = Includes.uploadManager
    val data: ArrayList<AttachmentEntry>
    private var textBody: String? = null
    private var currentPhotoCameraUri: Uri? = null
    private var timerValue: Long? = null
    fun getTimerValue(): Long? {
        return timerValue
    }

    fun setTimerValue(timerValue: Long?) {
        this.timerValue = timerValue
        resolveTimerInfoView()
    }

    fun resolveTimerInfoView() {
        view?.setTimerValue(timerValue)
    }

    private fun resolveTextView() {
        view?.setTextBody(textBody)
    }

    fun getTextBody(): String? {
        return textBody
    }

    fun setTextBody(body: String?) {
        textBody = body
        resolveTextView()
    }

    open val needParcelSavingEntries: ArrayList<AttachmentEntry>
        get() = ArrayList(0)

    override fun saveState(outState: Bundle) {
        super.saveState(outState)
        outState.putParcelable(SAVE_CURRENT_PHOTO_CAMERA_URI, currentPhotoCameraUri)
        outState.putParcelableArrayList(SAVE_DATA, needParcelSavingEntries)
        outState.putString(SAVE_BODY, textBody)
        if (timerValue != null) {
            outState.putLong(SAVE_TIMER, timerValue ?: return)
        }
    }

    fun onUploadProgressUpdate(updates: List<IProgressUpdate>) {
        for (update in updates) {
            val info = findInfoByPredicate(
                data
            ) {
                (it.attachment is Upload
                        && it.attachment.getObjectId() == update.id)
            }
            if (info != null) {
                val entry = info.second
                val obj = entry.attachment as Upload
                if (obj.status != Upload.STATUS_UPLOADING) {
                    continue
                }
                view?.updateProgressAtIndex(
                    entry.id,
                    update.progress
                )
            }
        }
    }

    fun onUploadObjectRemovedFromQueue(ids: IntArray) {
        for (id in ids) {
            val index = findUploadIndexById(id)
            if (index != -1) {
                manuallyRemoveElement(index)
            }
        }
    }

    fun onUploadQueueUpdates(updates: List<Upload>, predicate: (Upload) -> Boolean) {
        val startSize = data.size
        var count = 0
        for (u in updates) {
            if (predicate.invoke(u)) {
                data.add(AttachmentEntry(true, u))
                count++
            }
        }
        if (count > 0) {
            safelyNotifyItemsAdded(startSize, count)
        }
    }

    fun safelyNotifyItemsAdded(position: Int, count: Int) {
        view?.notifyItemRangeInsert(position, count)
    }

    fun combine(
        first: List<AttachmentEntry>,
        second: List<AttachmentEntry>
    ): List<AttachmentEntry> {
        val data: MutableList<AttachmentEntry> = ArrayList(safeCountOfMultiple(first, second))
        data.addAll(first)
        data.addAll(second)
        return data
    }

    fun onUploadStatusUpdate(update: Upload) {
        val index = findUploadIndexById(update.getObjectId())
        if (index != -1) {
            safeNotifyDataSetChanged()
        }
    }

    override fun onGuiCreated(viewHost: V) {
        super.onGuiCreated(viewHost)
        viewHost.displayInitialModels(data)
        resolveTimerInfoView()
        resolveTextView()
    }

    fun fireRemoveClick(index: Int, attachment: AttachmentEntry) {
        if (attachment.attachment is Upload) {
            val upload = attachment.attachment
            uploadManager.cancel(upload.getObjectId())
            return
        }
        onAttachmentRemoveClick(index, attachment)
    }

    fun safelyNotifyItemRemoved(position: Int) {
        view?.notifyItemRemoved(position)
    }

    open fun onAttachmentRemoveClick(index: Int, attachment: AttachmentEntry) {
        throw UnsupportedOperationException()
    }

    fun manuallyRemoveElement(index: Int) {
        data.removeAt(index)
        safelyNotifyItemRemoved(index)

        //safeNotifyDataSetChanged();
    }

    private val maxCountOfAttachments: Int
        get() = 10

    private fun canAttachMore(): Boolean {
        return data.size < maxCountOfAttachments
    }

    private val maxFutureAttachmentCount: Int
        get() {
            val count = data.size - maxCountOfAttachments
            return count.coerceAtLeast(0)
        }

    fun firePhotoFromVkChoose() {
        view?.openAddVkPhotosWindow(
            maxFutureAttachmentCount,
            accountId,
            accountId
        )
    }

    private fun checkAbilityToAttachMore(): Boolean {
        return if (canAttachMore()) {
            true
        } else {
            view?.showError(R.string.reached_maximum_count_of_attachments)
            false
        }
    }

    fun firePhotoFromLocalGalleryChoose() {
        if (!hasReadStoragePermission(applicationContext)) {
            view?.requestReadExternalStoragePermission()
            return
        }
        view?.openAddPhotoFromGalleryWindow(maxFutureAttachmentCount)
    }

    fun firePhotoFromCameraChoose() {
        if (!hasCameraPermission(applicationContext)) {
            view?.requestCameraPermission()
            return
        }
        createImageFromCamera()
    }

    private fun createImageFromCamera() {
        try {
            val photoFile = createImageFile()
            currentPhotoCameraUri = getExportedUriForFile(applicationContext, photoFile)
            view?.openCamera(currentPhotoCameraUri ?: return)
        } catch (e: IOException) {
            view?.showError(e.message)
        }
    }

    fun firePhotoMaked() {
        view?.notifySystemAboutNewPhoto(currentPhotoCameraUri ?: return)
        val makedPhoto = LocalPhoto().setFullImageUri(currentPhotoCameraUri)
        doUploadPhotos(listOf(makedPhoto))
    }

    protected open fun doUploadPhotos(photos: List<LocalPhoto>, size: Int) {
        throw UnsupportedOperationException()
    }

    protected open fun doUploadFile(file: String, size: Int) {
        throw UnsupportedOperationException()
    }

    fun fireFileSelected(file: String) {
        doUploadFile(file, Upload.IMAGE_SIZE_FULL)
    }

    private fun doFinalUploadPhotos(photos: List<LocalPhoto>, size: Int) {
        if (size == Upload.IMAGE_SIZE_CROPPING && photos.size == 1) {
            var to_up = photos[0].getFullImageUri()
            if (to_up?.path?.let { File(it).isFile } == true) {
                to_up = Uri.fromFile(to_up.path?.let { File(it) })
            }
            view?.displayCropPhotoDialog(to_up)
        } else {
            doUploadPhotos(photos, size)
        }
    }

    private fun doUploadPhotos(photos: List<LocalPhoto>) {
        val size = Settings.get()
            .main()
            .uploadImageSize
        if (size == null) {
            view?.displaySelectUploadPhotoSizeDialog(photos)
        } else {
            doFinalUploadPhotos(photos, size)
        }
    }

    fun firePhotosFromGallerySelected(photos: ArrayList<LocalPhoto>) {
        doUploadPhotos(photos)
    }

    fun fireButtonPhotoClick() {
        if (checkAbilityToAttachMore()) {
            view?.displayChoosePhotoTypeDialog()
        }
    }

    fun fireButtonAudioClick() {
        if (checkAbilityToAttachMore()) {
            view?.openAddAudiosWindow(
                maxFutureAttachmentCount,
                accountId
            )
        }
    }

    fun fireButtonVideoClick() {
        if (checkAbilityToAttachMore()) {
            view?.openAddVideosWindow(
                maxFutureAttachmentCount,
                accountId
            )
        }
    }

    fun fireButtonDocClick() {
        if (checkAbilityToAttachMore()) {
            view?.openAddDocumentsWindow(
                maxFutureAttachmentCount,
                accountId
            )
        }
    }

    protected open fun onPollCreateClick() {
        throw UnsupportedOperationException()
    }

    protected open fun onTimerClick() {
        throw UnsupportedOperationException()
    }

    fun fireButtonPollClick() {
        onPollCreateClick()
    }

    fun fireButtonTimerClick() {
        onTimerClick()
    }

    protected open fun onModelsAdded(models: List<AbsModel>) {
        for (model in models) {
            data.add(AttachmentEntry(true, model))
        }
        safeNotifyDataSetChanged()
    }

    fun fireAttachmentsSelected(attachments: ArrayList<out AbsModel>) {
        onModelsAdded(attachments)
    }

    fun fireUploadPhotoSizeSelected(photos: List<LocalPhoto>, size: Int) {
        doFinalUploadPhotos(photos, size)
    }

    fun firePollCreated(poll: Poll) {
        onModelsAdded(listOf(poll))
    }

    protected fun safeNotifyDataSetChanged() {
        view?.notifyDataSetChanged()
    }

    fun fireTextChanged(s: CharSequence?) {
        textBody = s.toString()
    }

    fun fireVkPhotosSelected(photos: ArrayList<Photo>) {
        onModelsAdded(photos)
    }

    fun fireCameraPermissionResolved() {
        if (hasCameraPermission(applicationContext)) {
            createImageFromCamera()
        }
    }

    fun fireReadStoragePermissionResolved() {
        if (hasReadStoragePermission(applicationContext)) {
            view?.openAddPhotoFromGalleryWindow(
                maxFutureAttachmentCount
            )
        }
    }

    fun hasUploads(): Boolean {
        for (entry in data) {
            if (entry.attachment is Upload) {
                return true
            }
        }
        return false
    }

    fun findUploadIndexById(id: Int): Int {
        for (i in data.indices) {
            val item = data[i]
            if (item.attachment is Upload && item.attachment.getObjectId() == id) {
                return i
            }
        }
        return -1
    }

    open fun fireTimerTimeSelected(unixtime: Long) {
        throw UnsupportedOperationException()
    }

    companion object {
        private const val SAVE_DATA = "save_data"
        private const val SAVE_TIMER = "save_timer"
        private const val SAVE_BODY = "save_body"
        private const val SAVE_CURRENT_PHOTO_CAMERA_URI = "save_current_photo_camera_uri"
        fun createFrom(objects: List<Upload>?): List<AttachmentEntry> {
            val data: MutableList<AttachmentEntry> = ArrayList(objects?.size ?: 0)
            for (obj in objects.orEmpty()) {
                data.add(AttachmentEntry(true, obj))
            }
            return data
        }


        fun createFrom(
            pairs: List<Pair<Int, AbsModel>>,
            canDelete: Boolean
        ): List<AttachmentEntry> {
            val data: MutableList<AttachmentEntry> = ArrayList(pairs.size)
            for (pair in pairs) {
                data.add(AttachmentEntry(canDelete, pair.second).setOptionalId(pair.first))
            }
            return data
        }
    }

    init {
        if (savedInstanceState != null) {
            currentPhotoCameraUri =
                savedInstanceState.getParcelableCompat(SAVE_CURRENT_PHOTO_CAMERA_URI)
            textBody = savedInstanceState.getString(SAVE_BODY)
            timerValue = if (savedInstanceState.containsKey(SAVE_TIMER)) savedInstanceState.getLong(
                SAVE_TIMER
            ) else null
        }
        data = ArrayList()
        if (savedInstanceState != null) {
            val savedEntries: ArrayList<AttachmentEntry>? =
                savedInstanceState.getParcelableArrayListCompat(
                    SAVE_DATA
                )
            if (savedEntries.nonNullNoEmpty()) {
                data.addAll(savedEntries)
            }
        }
    }
}